plugins/imkafka/imkafka.c

```diff
@@ -692,53 +707,53 @@ ENDfreeCnf
 /* This function is called to gather input.
  */
 BEGINrunInput
+	int i;
 	instanceConf_t *inst;
 CODESTARTrunInput
 	DBGPRINTF("imkafka: runInput loop started ...\n");
-	int activeListeners = 0;
+	activeKafkaworkers = 0;
 	for(inst = runModConf->root ; inst != NULL ; inst = inst->next) {
 		if(inst->rk != NULL) {
-			++activeListeners;
+			++activeKafkaworkers;
 		}
 	}
 
-	if(activeListeners == 0) {
+	if(activeKafkaworkers == 0) {
 		LogError(0, RS_RET_ERR, "imkafka: no active inputs, input does "
 			"not run - there should have been additional error "
 			"messages given previously");
 		ABORT_FINALIZE(RS_RET_ERR);
 	}
 
 
-	/* Start endless consumer loop - it is terminated when the thread is
-	 * signalled to do so. This, however, is handled by the framework.
-	 */
-	do {
-		for(inst = runModConf->root ; inst != NULL ; inst = inst->next) {
-			if(glbl.GetGlobalInputTermState() == 1)
-				break; /* terminate input! */
+	DBGPRINTF("imkafka: Starting %d imkafka workerthreads\n", activeKafkaworkers);
+	kafkaWrkrInfo = calloc(activeKafkaworkers, sizeof(struct kafkaWrkrInfo_s));
+	if (kafkaWrkrInfo == NULL) {
+		LogError(errno, RS_RET_OUT_OF_MEMORY, "imkafka: worker-info array allocation failed.");
+		ABORT_FINALIZE(RS_RET_OUT_OF_MEMORY);
+	}
 
-			if(inst->rk == NULL) {
-				continue;
-			}
+	/* Start worker threads for each imkafka input source
+	*/
+	i = 0;
+	for(inst = runModConf->root ; inst != NULL ; inst = inst->next) {
+		/* init worker info structure! */
+		kafkaWrkrInfo[i].inst = inst; /* Set reference pointer */
+		pthread_create(&kafkaWrkrInfo[i].tid, &wrkrThrdAttr, imkafkawrkr, &(kafkaWrkrInfo[i]));
+		i++;
+	}
 
-			// Try to add consumer only if connected! */
-			if(inst->bIsConnected == 1 && inst->bIsSubscribed == 0 ) {
-				addConsumer(runModConf, inst);
-			}
-			if(inst->bIsSubscribed == 1 ) {
-				msgConsume(inst);
-			}
-		}
+	while(glbl.GetGlobalInputTermState() == 0) {
 
 		/* Note: the additional 10000ns wait is vitally important. It guards rsyslog
 		 * against totally hogging the CPU if the users selects a polling interval
 		 * of 0 seconds. It doesn't hurt any other valid scenario. So do not remove.
-		 * rgerhards, 2008-02-14
 		 */
 		if(glbl.GetGlobalInputTermState() == 0)
 			srSleep(0, 100000);
-	} while(glbl.GetGlobalInputTermState() == 0);
+	}
+
+	// TODO: Wait for Worker threads to be finished!
```

Comment:
I think this comment no longer applies (see `afterRun`)? If so, please remove. Else we need to do the implementation.